src/java.base/share/classes/jdk/internal/io/BaseJdkConsoleImpl.java

```diff
@@ -0,0 +1,204 @@
+        return this;
+    }
+
+    @Override
+    public String readln(String prompt) {
+        String line = null;
+        synchronized (writeLock) {
+            synchronized(readLock) {
+                pw.print(prompt);
+                pw.flush(); // automatic flushing does not cover print
+                try {
+                    char[] ca = readline(false);
+                    if (ca != null)
+                        line = new String(ca);
+                } catch (IOException x) {
+                    throw new IOError(x);
+                }
+            }
+        }
+        return line;
+    }
+
+    @Override
+    public String readln() {
+        String line = null;
+        synchronized(readLock) {
+            try {
+                char[] ca = readline(false);
+                if (ca != null)
+                    line = new String(ca);
+            } catch (IOException x) {
+                throw new IOError(x);
+            }
+        }
+        return line;
+    }
+
+    @Override
+    public JdkConsole format(Locale locale, String format, Object ... args) {
+        formatter.format(locale, format, args).flush();
+        return this;
+    }
+
+    @Override
+    public String readLine(Locale locale, String format, Object ... args) {
+        String line = null;
+        synchronized (writeLock) {
+            synchronized(readLock) {
+                if (!format.isEmpty())
+                    pw.format(locale, format, args);
+                try {
+                    char[] ca = readline(false);
+                    if (ca != null)
+                        line = new String(ca);
+                } catch (IOException x) {
+                    throw new IOError(x);
+                }
+            }
+        }
+        return line;
+    }
+
+    @Override
+    public String readLine() {
+        return readLine(Locale.getDefault(Locale.Category.FORMAT), "");
+    }
+
+    @Override
+    public char[] readPassword(Locale locale, String format, Object ... args) {
+        char[] passwd = null;
+        synchronized (writeLock) {
+            synchronized(readLock) {
+                try {
+                    if (!format.isEmpty())
+                        pw.format(locale, format, args);
+                    passwd = readline(true);
+                } catch (IOException x) {
+                    throw new IOError(x);
+                }
+                pw.println();
+            }
+        }
+        return passwd;
+    }
+
+    @Override
+    public char[] readPassword() {
+        return readPassword(Locale.getDefault(Locale.Category.FORMAT), "");
+    }
+
+    @Override
+    public void flush() {
+        pw.flush();
+    }
+
+    @Override
+    public Charset charset() {
+        return charset;
+    }
+
+    protected Reader wrapReader(Reader baseReader) {
+        return baseReader;
+    }
+
+    protected final Charset charset;
+    protected final Object readLock;
+    protected final Object writeLock;
```

Comment:
Recommend mentioning that when both locks must be acquired, the `writeLock` must be acquired first. Unfortunately we cannot just use a `ReentrantReadWriteLock`.